dora/core/client/fs/src/main/java/alluxio/client/file/DoraFileOutStreamBak.java

```diff
@@ -0,0 +1,347 @@
+      mBlockSize = options.getBlockSizeBytes();
+      mAlluxioStorageType = options.getAlluxioStorageType();
+      mUnderStorageType = options.getUnderStorageType();
+      mOptions = options;
+      mBlockStore = BlockStoreClient.create(mContext);
+      mPreviousBlockOutStreams = new ArrayList<>();
+      mClosed = false;
+      mCanceled = false;
+      mShouldCacheCurrentBlock = mAlluxioStorageType.isStore();
+      mBytesWritten = 0;
+
+      if (!mUnderStorageType.isSyncPersist()) {
+        mUnderStorageOutputStream = null;
+      } else { // Write is through to the under storage, create mUnderStorageOutputStream.
+        // Create retry policy for initializing write.
+        AlluxioConfiguration pathConf = mContext.getPathConf(path);
+        RetryPolicy initRetryPolicy = ExponentialTimeBoundedRetry.builder()
+            .withMaxDuration(pathConf.getDuration(PropertyKey.USER_FILE_WRITE_INIT_MAX_DURATION))
+            .withInitialSleep(pathConf.getDuration(PropertyKey.USER_FILE_WRITE_INIT_SLEEP_MIN))
+            .withMaxSleep(pathConf.getDuration(PropertyKey.USER_FILE_WRITE_INIT_SLEEP_MAX))
+            .withSkipInitialSleep().build();
+        // Try find a worker from policy.
+        Optional<WorkerNetAddress> workerNetAddress = Optional.empty();
+        while (!workerNetAddress.isPresent() && initRetryPolicy.attempt()) {
+          GetWorkerOptions getWorkerOptions = GetWorkerOptions.defaults()
+                  .setBlockWorkerInfos(mContext.getCachedWorkers())
+                  .setBlockInfo(new BlockInfo()
+                  .setBlockId(-1)
+                  .setLength(0)); // not storing data to Alluxio, so block size is 0
+          workerNetAddress = options.getLocationPolicy().getWorker(getWorkerOptions);
+        }
+        if (!workerNetAddress.isPresent()) {
+          // Assume no worker is available because block size is 0.
+          throw new UnavailableException(ExceptionMessage.NO_WORKER_AVAILABLE.getMessage());
+        }
+        mUnderStorageOutputStream = mCloser
+            .register(UnderFileSystemFileOutStream.create(mContext,
+                workerNetAddress.get(), mOptions));
+      }
+    } catch (Throwable t) {
+      throw CommonUtils.closeAndRethrow(mCloser, t);
+    }
+  }
+
+  @Override
+  public void cancel() throws IOException {
+    mCanceled = true;
+    close();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mClosed) {
+      return;
+    }
+    try (Timer.Context ctx = MetricsSystem
+            .uniformTimer(MetricKey.CLOSE_ALLUXIO_OUTSTREAM_LATENCY.getName()).time()) {
+      if (mCurrentBlockOutStream != null) {
+        mPreviousBlockOutStreams.add(mCurrentBlockOutStream);
+      }
+
+      CompleteFilePOptions.Builder optionsBuilder = CompleteFilePOptions.newBuilder();
+      optionsBuilder.setCommonOptions(FileSystemMasterCommonPOptions.newBuilder()
+          .setOperationId(new OperationId(UUID.randomUUID()).toFsProto()).buildPartial());
```

Comment:
The UUID here is not really useful anymore?

This operation ID was introduced in https://github.com/Alluxio/alluxio/pull/14162
Every operation has this UUID and the Alluxio master keeps a Set<UUID>. The UUID is also written into journal and therefore known by the standby masters. So if the primary has processed the CompleteFile/CreateFile, but crashes before responding to the client, the client will retry sending the same request (now answered by the standby, now primary), and the new primary will know "Oh this request is actually completed and therefore no need to handle". 

Imagine in 2.x what happens IF we DO NOT have this operation ID:
1. Client sends CreateFile to primary
2. Primary creates this Inode and writes journal, journal propagates to standby
3. Before primary answers client, primary crashes
4. A standby becomes primary, of course the standby DOES NOT have this Set<UUID>
5. Client resends CreateFile
6. New primary actually sees this existing Inode, and rejects with FILE_ALREADY_EXISTS

Above tells you why we want to tell one request is actually duplicated. We should skip it instead of returning an incorrect error.

Okay back to this new code. There is no master inode journal and do you still face the same situation? What happens if your worker completes the file and dies before returning to the client? Do you want a new worker to be able to tell the client retry is actually a duplicate? I don't think we can achieve that in the existing architecture. 

Thus I think you don't need this UUID here. Especially UUID is SLOW! I think it is fine to just leave an empty Operation ID somehow? And leave a comment here or create a jira ticket. We shall revisit the operation ID when we implement fault tolerance to dora.

Reply:
You're right. We do have a UUID to handle the similar logic in `DoraCacheFileSystem#createFile`. We can improve the performance by replacing the UUID in other way. I left a `TODO` in the code.